### PR TITLE
CONSOLE-2376: Dynamic plugins for Console

### PR DESCRIPTION
**Pull request by jhadvig**
API changes that are mapping the proposed [enhancement](https://github.com/openshift/enhancements/pull/441/files) doc. The enhancement is still not merged so creating a first draft of it's API changes

Will generate the crd yaml once we align on the API changes.

/assign @spadgett @bparees 

https://issues.redhat.com/browse/CONSOLE-2377

### CODE REVIEW

**Comment by spadgett** [comment]
@jhadvig In addition to the operator config, we'll need the `ConsolePlugin` CRD in the `console.openshift.io` API group.

**Comment by jhadvig** [comment]
@spadgett only added the `ConsolePlugin` type, for now... we can generate the CRD after we align on the working and the struct I guess ?

File: operator/v1/types_console.go
Comment (spadgett) [inline operator/v1/types_console.go:47]: The operator config only has the plugin names, which reference a `ConsolePlugin` resource. The `ConsolePlugin` has the other details of the operator like the service.

**Review by spadgett — APPROVED** [review]
/lgtm

File: console/v1/types_console_plugin.go
Comment (spadgett) [inline console/v1/types_console_plugin.go:10]:
```suggestion
// dynamically loading code from another service running on the cluster.
```

**Comment by sferich888** [comment]
What Jira is this connected to?

**Comment by spadgett** [comment]
@sferich888 It's https://issues.redhat.com/browse/CONSOLE-2377

**Comment by jhadvig** [comment]
@spadgett I've generated the CRDs and addressed your comment, which removed the `lgtm` label. PTAL

File: console/v1/types_console_plugin.go
Comment (mfojtik) [inline console/v1/types_console_plugin.go:26]: nit: `ServiceName` or `SecureServiceName`

File: console/v1/types_console_plugin.go
Comment (vojtechszocs) [inline console/v1/types_console_plugin.go:40]: The name of Console dynamic plugin manifest file (`plugin-manifest.json`) is fixed.

In the above data structure, `name` + `namespace` + `port` are used to resolve the plugin service URL address. Console web application refers to this address as "plugin base URL".

Given the above, I think that the `manifest` field is redundant.

**Review by vojtechszocs — CHANGES_REQUESTED** [review]

File: console/v1/types_console_plugin.go
Comment (spadgett) [reply console/v1/types_console_plugin.go:40]: Is there a reason to force a specific path for the manifest file, particularly when it's so easy to make it a property? You might want to serve your plugin from a service that does other things as well and move the plugin assets under a specific directory as an example.

Comment (spadgett) [reply console/v1/types_console_plugin.go:40]: I talked to @vojtechszocs. I'm good changing this to something like `basePath` and having the manifest always be at `<basePath>/plugin-manifest.json`. But either way we need some property here to let plugins have a context root for the plugin and not assume everything can be served at the base path.

File: operator/v1/types_console.go
Comment (vojtechszocs) [reply operator/v1/types_console.go:47]: > Console operator should ensure that _currently enabled plugins list is a subset of all plugins_

@spadgett @christianvogt Given the possibility of race between `ConsolePlugin` resource being added (after operator install) vs. Console operator config being updated (checkbox that enables a dynamic plugin as part of operator install flow), the UI can deal with non-existing plugins.

When some `plugin-manifest.json` file(s) cannot be loaded, we can show a visual notification.

Comment (vojtechszocs) [reply operator/v1/types_console.go:47]: I see that in openshift/console-operator#478 the ConfigMap uses `plugins` field so I'm OK with keeping this field named `plugins` as well, keeping the naming consistent.

File: operator/v1/types_console.go
Comment (vojtechszocs) [reply operator/v1/types_console.go:45]: Please scratch my comment above if `metadata.name` must be unique across all `ConsolePlugin` custom resources.

File: console/v1/types_console_plugin.go
Comment (vojtechszocs) [reply console/v1/types_console_plugin.go:40]:
```suggestion
	BasePath string `json:"manifest"`
```

File: operator/v1/types_console.go
Comment (jhadvig) [reply operator/v1/types_console.go:45]: > I think we should implement an equivalent constraint when installing an operator, so that we don't end up with duplicates in the "enabled plugins" list.

 Yes the logic that prevents duplicate plugin names is in the operator ATM. `removeUnavailablePlugins` removes also dups. Will add it into the function comments.

File: console/v1/types_console_plugin.go
Comment (jhadvig) [reply console/v1/types_console_plugin.go:40]: > But either way we need some property here to let plugins have a context root for the plugin and not assume everything can be served at the base path.

Not sure why that can't be handled by the `BasePath` ?
Wont the plugin server should be handling everything on the `<endpoint>:<port>/<basepath>/*` ?

File: console/v1/0000_10_consoleplugin.crd.yaml
Comment (jhadvig) [inline console/v1/0000_10_consoleplugin.crd.yaml:14]: Thinking if we should not put this to `v1alpha1`, in case we decide to change any fields for 4.8
@spadgett thoughts ?

Comment (spadgett) [reply console/v1/0000_10_consoleplugin.crd.yaml:14]: We did say it would be alpha in the enhancement since it's only meant for internal use at this time.

**Comment by jhadvig** [comment]
There have been small adjustments to the API change. PTAL

**Comment by yapei** [comment]
LGTM from QE's point of view

**Comment by spadgett** [comment]
> LGTM from QE's point of view

Adding qe-approved label based on this comment and QE approval of https://github.com/openshift/console/pull/7360 which included the API change.

/label qe-approved

**Comment by spadgett** [comment]
Adding the docs approved label now that it's available in the repo (see @ahardin-rh approval above).

/label docs-approved

**Comment by jhadvig** [comment]
@reestr PTAL

**Review by spadgett — APPROVED** [review]
/lgtm
/hold cancel

Console and console operator PRs have been tested by QE and have all approvals.

**Comment by spadgett** [comment]
Manually adding the `bugzilla/valid-bug` label since the console team is piloting the new [feature process that does not require a Bugzilla](https://github.com/openshift/console/pull/7360#issuecomment-745419809).

**Pull request merged**
